### PR TITLE
feat(thumbnails): add support for user specific thumbs

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
This PR adds support for generating user-specific thumbnails. This is a typical requirement in environments where some form of user impersonation is being used, and sharing thumbnails across all users with access to the same dashboards/charts could leak sensitive data.

This PR does the following:
- Renames `ReportScheduleExecutor` to `ExecutorType` so that it can be reused for thumbnails. Also moves the utils from the `reports` package to the `tasks` package, as it's shared with thumbnails now.
- Adds a new package `thumbnails` to contain the thumbs-specific types etc. Also move the thumbnail task module here and deprecate the old one (it will still work, but will now emit a deprecation warning).
- Adds a new executor type called `CURRENT_USER` which corresponds to the logged-in user that initiated the request. This user will be undefined for Alerts & Reports (=Celery has initiated those), but for thumbnails, this is the user that requested the thumbnail.
- Adds the following config options:
  - `THUMBNAIL_EXECUTE_AS` - similar config as for Alerts & Reports
  - `THUMBNAIL_DASHBOARD_DIGEST_FUNC`: callback for generating custom digests for dashboards. This is handy if a deployment wants to use different hashing functions or use advanced logic for deciding if a thumbnail can be shared across a larger user pool or not.
  - `THUMBNAIL_CHART_DIGEST_FUNC`: callback for generating custom digests for charts

By default the digests will stay unchanged, as the new default value for `THUMBNAIL_EXECUTE_AS = [ExecutorType.SELENIUM]`. However, when setting it to `[ExecutorType.CURRENT_USER]`, the username will be added to the `unique_string` prior to hashing to make it unique per user.

### AFTER
This chart thumbnail was cached using the following query on a Trino database connection using user impersonation with the following virtual dataset:
```sql
select concat('Database: ', current_user) as user, 1 as num union all
select 'Jinja: {{ current_username() }}' as user, 2 as num
```
image

As can be seen, both the `current_user` (rendered by Trino) and `{{ current_username() }}` (rendered by Superset) both show the user as being `v_brofeldt`, i.e. not a service account.

### BEFORE
When I changed the chart to reference a postgres database with basic auth using the username `postgres` and `THUMBNAIL_EXECUTE_AS = [ExecutorType.SELENIUM]` (=same as current behavior), the result was as follows:
image

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/models/slice.py
Comment (villebro) [inline superset/models/slice.py:348]: There is a similar method in `Dashboard` that bypasses the base filter

File: superset/tasks/utils.py
Comment (villebro) [inline superset/tasks/utils.py:38]: The reason why this function returns the username as a string, and not the `User` object, is because this function will be called frequently (e.g. when getting all charts/dashboards). Since we only know the selenium username in the config, we would otherwise have to fetch it from the metastore, causing unnecessary round trips to the metastore.

File: superset/thumbnails/tasks.py
Comment (villebro) [inline superset/thumbnails/tasks.py:1]: This file is really the same as `superset/tasks/thumbnails.py` before, just updated with the new logic (fetching the executor and overriding the username etc)

File: tests/integration_tests/reports/alert_tests.py
Comment (villebro) [inline tests/integration_tests/reports/alert_tests.py:55]: this is really just one added test for the new `INITIATOR` case, other than that it's just updating the test cases to conform to the new sig of `get_executor`

**Comment by kamalkeshavani-aiinside** [comment]
@villebro Thank you for implementation of this feature. Just sharing our usecase.
In our Superset, we keep the dashboard thumbnails cached everyday for all users for better user experience. We do plan to create a dashboard with user specific content, where we can use this feature of each user generating their own thumbnails. 

But this implementation will force the thumbnail generation for all the dashboards for each user, which is not ideal for us.
An ideal solution for our usecase would be control of executor for each dashboard/chart, so the owner of dashboard/chart can decide if that particular dashboard/chart needs `INITIATOR` as executor for thumbnail. Whether it is actually feasible or not, I will let you decide.

**Comment by villebro** [comment]
> Just sharing our usecase. In our Superset, we keep the dashboard thumbnails cached everyday for all users for better user experience. We do plan to create a dashboard with user specific content, where we can use this feature of each user generating their own thumbnails.
> 
> But this implementation will force the thumbnail generation for all the dashboards for each user, which is not ideal for us. An ideal solution for our usecase would be control of executor for each dashboard/chart, so the owner of dashboard/chart can decide if that particular dashboard/chart needs `INITIATOR` as executor for thumbnail. Whether it is actually feasible or not, I will let you decide.

@kamalkeshavani-aiinside we can certainly consider adding this in a future PR. All it would really require is adding a `thumbnail_executor` field in the Dashboard and Slice models and then add a dropdown in their respective modals (if undefined, it would default to the global config).

Would you be open to working on this feature? I'm happy to provide guidance and review help if needed.

File: superset/tasks/types.py
Comment (villebro) [inline superset/tasks/types.py:21]: This is the same type as the previous `ReportScheduleExecutor`, but with the added `INITIATOR` enum.

File: superset/dashboards/api.py
Comment (dpgaspar) [inline superset/dashboards/api.py:883]: here, note that the `self._base_filters` provides our security against users being able to view each others thumbnails/screenshots

File: superset/models/slice.py
Comment (dpgaspar) [inline superset/models/slice.py:237]: don't see any other uses of `get_chart_digest` why not just post it's contents here?

File: superset/thumbnails/digest.py
Comment (dpgaspar) [inline superset/thumbnails/digest.py:40]: If the API endpoints for screenshots provide enough security, we probably don't need this function anymore.

File: docs/docs/installation/cache.mdx
Comment (dpgaspar) [inline docs/docs/installation/cache.mdx:60]: nit: at first I struggled with the name `INITIATOR`, after reading the entire PR I got more used to it :)
still do you think `ExecutorType.CURRENT_USER` or similar could be more explicit?

Comment (villebro) [reply docs/docs/installation/cache.mdx:60]: I originally had `CURRENT_USER`, but for some reason I decided `INITIATOR` was better. But now that you said it, yeah, `CURRENT_USER` makes much more sense. I'll update it.

File: superset/models/slice.py
Comment (villebro) [reply superset/models/slice.py:346]: I just copied this method from the dashboard class, let me re-evaluate how this should be done.

File: superset/tasks/utils.py
Comment (villebro) [reply superset/tasks/utils.py:90]: oops!

File: superset/dashboards/api.py
Comment (villebro) [reply superset/dashboards/api.py:883]: This is a VERY good point that I hadn't thought of! Thanks for catching this!

File: superset/models/slice.py
Comment (villebro) [reply superset/models/slice.py:237]: Good idea 👍

File: superset/thumbnails/digest.py
Comment (villebro) [reply superset/thumbnails/digest.py:40]: I agree, I'll simplify this.

File: superset/models/dashboard.py
Comment (villebro) [inline superset/models/dashboard.py:441]: bycatch

File: tests/integration_tests/thumbnails_tests.py
Comment (villebro) [inline tests/integration_tests/thumbnails_tests.py:156]: Since the digest may now be affected by who is logged in, all tests are updated to fetch the thumbnail URL via the API after login.

**Comment by kamalkeshavani-aiinside** [comment]
> @kamalkeshavani-aiinside we can certainly consider adding this in a future PR. All it would really require is adding a thumbnail_executor field in the Dashboard and Slice models and then add a dropdown in their respective modals (if undefined, it would default to the global config).
> 
> Would you be open to working on this feature? I'm happy to provide guidance and review help if needed.
Thank you for the suggestion. Sure, I can try to work on this with your help.

File: superset/thumbnails/digest.py
Comment (dpgaspar) [inline superset/thumbnails/digest.py:68]: our current dashboard digest is:
``` python
    @property
    def digest(self) -> str:
        """
        Returns a MD5 HEX digest that makes this dashboard unique
        """
        unique_string = f"{self.position_json}.{self.css}.{self.json_metadata}"
        return md5_sha_from_str(unique_string)
```

Adding an executor will invalidate all current computed dashboard thumbnails.

I would also prefer to bring these changes back to `/superset/tasks/thumbnails` and avoid introducing the deprecation and underlying breaking change on the task structure.

Comment (villebro) [reply superset/thumbnails/digest.py:68]: > Adding an executor will invalidate all current computed dashboard thumbnails.

Oof, that was an unintended mistake, the executor was only supposed to be added in the case of `CURRENT_USER`.

> I would also prefer to bring these changes back to /superset/tasks/thumbnails and avoid introducing the deprecation and underlying breaking change on the task structure.

Makes sense - I'll revert the move

**Review by AAfghahi — APPROVED** [review]

**Review by michael-s-molina — APPROVED** [review]
Code LGMT. 

@villebro If you don't mind, it would be a good idea to also get @dpgaspar's approval before merging it. Thank you for the improvement!

**Comment by villebro** [comment]
> Code LGMT. 
> 
> 
> 
> @villebro If you don't mind, it would be a good idea to also get @dpgaspar's approval before merging it. Thank you for the improvement!

Absolutely @michael-s-molina 👍 As I also found a bug today in the PR I'm going to let it simmer over the weekend as I feel there's room for improvement in the tests.

**Review by dpgaspar — APPROVED** [review]
Looks really good, since this change will invalidate thumbnails cache it would be nice to add a note on UPDATING.md

**Pull request merged**
